kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h

```diff
@@ -0,0 +1,1182 @@
+     * @param rModelPart The model part of the problem to solve
+     * @param rb The RHS vector
+     */
+    void BuildRHS(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemVectorType& rb
+        ) override
+    {
+        KRATOS_TRY
+
+        // Resize again the system to the original size
+        if (rb.size() != BaseType::mEquationSystemSize) {
+            rb.resize(BaseType::mEquationSystemSize, false);
+        }
+
+        // Build the base RHS
+        BaseType::BuildRHS(pScheme, rModelPart, rb);
+
+        KRATOS_CATCH("")
+    }
+
+    /**
+     * @brief This method computes the reactions of the system
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rA The LHS matrix
+     * @param rDx The Unknowns vector
+     * @param rb The RHS vector
+     */
+    void CalculateReactions(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb
+        ) override
+    {
+        TSparseSpace::SetToZero(rb);
+
+        // Refresh RHS to have the correct reactions
+        BaseType::BuildRHSNoDirichlet(pScheme, rModelPart, rb);
+
+        const int ndofs = static_cast<int>(BaseType::mDofSet.size());
+
+        // First iterator
+        const auto it_dof_begin = BaseType::mDofSet.begin();
+
+        //NOTE: dofs are assumed to be numbered consecutively in the BlockBuilderAndSolver
+        #pragma omp parallel for
+        for (int k = 0; k<ndofs; k++) {
+            auto it_dof =  it_dof_begin + k;
+
+            if (it_dof->IsFixed()) {
+                it_dof->GetSolutionStepReactionValue() = -rb[it_dof->EquationId()];
+            }
+        }
+
+        // NOTE: The constraints reactions are already computed when solving the dofs
+        const int number_slave_dofs = BaseType::mSlaveIds.size();
+        #pragma omp parallel for
+        for (int k = 0; k<number_slave_dofs; k++) {
+            const IndexType equation_id = BaseType::mSlaveIds[k];
+            auto it_dof = it_dof_begin + equation_id;
+            it_dof->GetSolutionStepReactionValue() = mLagrangeMultiplierVector[mCorrespondanceDofsSlave[equation_id]];
+        }
+
+    }
+
+    /**
+     * @brief Applies the dirichlet conditions. This operation may be very heavy or completely
+     * unexpensive depending on the implementation choosen and on how the System Matrix is built.
+     * @details For explanation of how it works for a particular implementation the user
+     * should refer to the particular Builder And Solver choosen
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     * @param rA The LHS matrix
+     * @param rDx The Unknowns vector
+     * @param rb The RHS vector
+     */
+    void ApplyDirichletConditions(
```

Comment:
i also think this function is identical to the base class

Reply:
In some point, it wasn't, removing

Reply:
Nope, they are different, otherwise tests fail. The difference is:

~~~cpp
// Filling with ones the LM dofs
        #pragma omp parallel for firstprivate(ndofs)
        for (int k = ndofs; k<static_cast<int>(system_size); ++k) {
            scaling_factors[k] = 1.0;
        }
~~~